SVS Vamana LeanVec: Cannot serialize trained-only index

## Summary

 The issue is that `IndexSVSVamanaLeanVec` can't be serialized (or deserialized) after training but before adding vectors. This breaks the standard model-based training workflow that works fine for IVF and HNSW-PQ.

**Error:** `"Cannot serialize: SVS index not initialized"`

## Background: How OpenSearch k-NN Uses Model-Based Training

OpenSearch k-NN has a model-based training API that lets users train once and reuse the trained parameters across multiple indexes. Here's how it works:

### The Workflow

```
1. User calls: POST /_plugins/_knn/models/_train
   - OpenSearch creates empty index via index_factory()
   - Calls train() with training data
   - Serializes trained model (works for IVF/HNSW-PQ)
   - Saves to internal .opensearch-knn-models index
   
2. User creates index: PUT /my-index (with "model_id": "my-trained-model")
   - OpenSearch loads the serialized trained model (works for IVF/HNSW-PQ)
   - Creates production index from that model
   
3. User adds documents: POST /my-index/_bulk
   - Vectors get added to the production index
   - Multiple indexes can share the same trained model
```



### The Code Path (OpenSearch side)

```cpp
// Step 1: Train Model API - faiss_wrapper.cpp:TrainIndex()
indexWriter = index_factory(description);  // e.g., "IVF256,PQ8"

if (!indexWriter->is_trained) {
    indexWriter->train(n, trainingData);  // Compute quantizers/codebooks
}

faiss::write_index(indexWriter.get(), &writer);  // Works for IVF/HNSW-PQ
                                                  // Fails for SVS LeanVec

// Step 2: Create Production Index API - faiss_wrapper.cpp:CreateIndexFromTemplate()
indexWriter.reset(faiss::read_index(&reader));  // Load trained model
idMap.add_with_ids(n, vectors, ids);             // Add vectors
faiss::write_index(&idMap, &writer);             // Save complete index
```



## Reproduction

Here's a minimal test that shows the problem:

### Test Code

```cpp
#include <faiss/Index.h>
#include <faiss/index_factory.h>
#include <faiss/index_io.h>
#include <faiss/impl/io.h>
#include <vector>
#include <random>

std::vector<float> generate_vectors(int dim, int n) {
    std::vector<float> data(dim * n);
    std::mt19937 gen(42);
    std::uniform_real_distribution<float> dis(0.0f, 1.0f);
    for (auto& x : data) x = dis(gen);
    return data;
}

int main() {
    int dim = 128;
    int n_train = 10000;
    auto train_data = generate_vectors(dim, n_train);
    
    // Test 1: IVF256,PQ8 (Baseline - Should Work)
    std::cout << "Test 1: IVF256,PQ8" << std::endl;
    faiss::Index* ivf = faiss::index_factory(dim, "IVF256,PQ8", faiss::METRIC_L2);
    ivf->train(n_train, train_data.data());
    
    faiss::VectorIOWriter writer1;
    faiss::write_index(ivf, &writer1);
    std::cout << "  Serialized " << writer1.data.size() << " bytes\n" << std::endl;
    delete ivf;
    
    // Test 2: SVS Vamana + LeanVec (This is where it fails)
    std::cout << "Test 2: SVSVamana64,LeanVec4x4_16" << std::endl;
    faiss::Index* svs = faiss::index_factory(dim, "SVSVamana64,LeanVec4x4_16", faiss::METRIC_L2);
    svs->train(n_train, train_data.data());
    std::cout << "  is_trained: " << svs->is_trained << std::endl;
    
    try {
        faiss::VectorIOWriter writer2;
        faiss::write_index(svs, &writer2);
        std::cout << "  Serialized " << writer2.data.size() << " bytes" << std::endl;
    } catch (const std::exception& e) {
        std::cout << "  FAILED: " << e.what() << std::endl;
    }
    delete svs;
    
    return 0;
}
```


### What You'll See

```
Test 1: IVF256,PQ8
  Serialized 262324 bytes

Test 2: SVSVamana64,LeanVec4x4_16
  is_trained: 1
  FAILED: Error in void faiss::IndexSVSVamana::serialize_impl(std::ostream&) const 
          at IndexSVSVamana.cpp:465: Error: '!(impl)' failed: 
          Cannot serialize: SVS index not initialized.
```


## What We Need

For OpenSearch k-NN to work with LeanVec, we need the ability to serialize and deserialize the trained state without vectors. Specifically:

### 1. Serialize Trained Parameters

When `is_trained == true` but `impl == null`.

### 2. Deserialize Trained Parameters

When loading a trained-only index (no vectors yet), we need to restore trained parameters and leave `impl` as null.

--- a/faiss/impl/index_read.cpp
+++ b/faiss/impl/index_read.cpp
@@ -1303,22 +1303,29 @@ Index* read_index(IOReader* f, int io_flags) {
             READ1(dynamic_cast<IndexSVSVamanaLeanVec*>(svs)->leanvec_d);
             READ1(dynamic_cast<IndexSVSVamanaLeanVec*>(svs)->leanvec_level);
         }
-
-        // Create a non-buffering stream adapter from IOReader
-        faiss::svs_io::ReaderStreambuf rbuf(f);
-        std::istream is(&rbuf);
-        svs->deserialize_impl(is);
-        idx = svs;
+        bool initialized;
+        READ1(initialized);
+        if (initialized) {
+            // Create a non-buffering stream adapter from IOReader
+            faiss::svs_io::ReaderStreambuf rbuf(f);
+            std::istream is(&rbuf);
+            svs->deserialize_impl(is);
+            idx = svs;
+        }
     } else if (h == fourcc("ISVF")) {
         // SVS Flat
         IndexSVSFlat* svs = new IndexSVSFlat();
         read_index_header(svs, f);
 
-        // Create a non-buffering stream adapter from IOReader
-        faiss::svs_io::ReaderStreambuf rbuf(f);
-        std::istream is(&rbuf);
-        svs->deserialize_impl(is);
-        idx = svs;
+        bool initialized;
+        READ1(initialized);
+        if (initialized) {
+            // Create a non-buffering stream adapter from IOReader
+            faiss::svs_io::ReaderStreambuf rbuf(f);
+            std::istream is(&rbuf);
+            svs->deserialize_impl(is);
+            idx = svs;
+        }
     }
 #endif // FAISS_ENABLE_SVS
     else {

--- a/faiss/impl/index_write.cpp
+++ b/faiss/impl/index_write.cpp
@@ -947,24 +947,32 @@ void write_index(const Index* idx, IOWriter* f, int io_flags) {
             WRITE1(lean->leanvec_level);
         }
 
-        // Wrap SVS I/O and stream to IOWriter
-        faiss::BufferedIOWriter bwr(f);
-        faiss::svs_io::WriterStreambuf wbuf(&bwr);
-        std::ostream os(&wbuf);
-        svs->serialize_impl(os);
-        os.flush();
+        bool initialized = (svs->impl != nullptr);
+        WRITE1(initialized);
+        if (initialized) {
+            // Wrap SVS I/O and stream to IOWriter
+            faiss::BufferedIOWriter bwr(f);
+            faiss::svs_io::WriterStreambuf wbuf(&bwr);
+            std::ostream os(&wbuf);
+            svs->serialize_impl(os);
+            os.flush();
+        }
     } else if (
             const IndexSVSFlat* svs = dynamic_cast<const IndexSVSFlat*>(idx)) {
         uint32_t h = fourcc("ISVF");
         WRITE1(h);
         write_index_header(idx, f);
 
-        // Wrap SVS I/O and stream to IOWriter
-        faiss::BufferedIOWriter bwr(f);
-        faiss::svs_io::WriterStreambuf wbuf(&bwr);
-        std::ostream os(&wbuf);
-        svs->serialize_impl(os);
-        os.flush();
+        bool initialized = (svs->impl != nullptr);
+        WRITE1(initialized);
+        if (initialized) {
+            // Wrap SVS I/O and stream to IOWriter
+            faiss::BufferedIOWriter bwr(f);
+            faiss::svs_io::WriterStreambuf wbuf(&bwr);
+            std::ostream os(&wbuf);
+            svs->serialize_impl(os);
+            os.flush();
+        }
     }
 #endif // FAISS_ENABLE_SVS
     else {

